lib/rules/no-object-constructor.js

```diff
@@ -80,6 +81,74 @@ module.exports = {
             return false;
         }
 
+        /**
+         * Determines whether a parenthesized object literal that replaces a specified node needs to be preceded by a semicolon.
+         * @param {ASTNode} node The node to be replaced.
+         * @returns {boolean} Whether a semicolon is required before the parenthesized object literal.
+         */
+        function needsSemicolon(node) {
+            const prevToken = sourceCode.getTokenBefore(node);
+
+            if (
+                !prevToken ||
+                prevToken.type === "Punctuator" && [":", ";", ">", "{", "=>", "++", "--"].includes(prevToken.value)
+            ) {
+                return false;
+            }
+
+            const prevNode = sourceCode.getNodeByRangeIndex(prevToken.range[0]);
+
+            if (isClosingParenToken(prevToken)) {
+                return ![
+                    "DoWhileStatement",
+                    "ForInStatement",
+                    "ForOfStatement",
+                    "ForStatement",
+                    "IfStatement",
+                    "WhileStatement",
+                    "WithStatement"
+                ].includes(prevNode.type);
+            }
+
+            if (isClosingBraceToken(prevToken)) {
+                return (
+                    prevNode.type === "BlockStatement" && prevNode.parent.type === "FunctionExpression" ||
+                    prevNode.type === "ClassBody" && prevNode.parent.type === "ClassExpression" ||
+                    prevNode.type === "ObjectExpression"
+                );
+            }
+
+            if (prevToken.type === "Identifier") {
+                return !["BreakStatement", "ContinueStatement"].includes(prevNode.parent.type);
+            }
```

Comment:
An edge case:

```js
// sourceType: "script", default parser (espree)

var yield = 5;

yield: while (foo) {
    if (bar)
        break yield
    new Object(); // suggestion adds unnecessary `;` because `yield` is a "Keyword" token
}

```

Whether a token will end up being keyword or identifier is generally unreliable. For example, `yield` is always Keyword for espree but always Identifier for `@typescript-eslint/parser`. It might be best to merge checks related to identifiers and checks related to keywords under `if (prevToken.type === "Identifier" || prevToken.type === "Keyword")`.

Reply:
Good catch! BTW, the autofix logic will clearly not always work well in TypeScript, as it will sometimes insert an unnecessary semicolon:

```ts
type a = b
Object();
```